projects/cds/src/merchandising/cms-components/merchandising-carousel/merchandising-carousel.component.ts

```diff
@@ -24,19 +26,81 @@ export class MerchandisingCarouselComponent {
   title$: Observable<string> = this.componentData$.pipe(
     map(data => data.title)
   );
-  items$: Observable<Observable<Product>[]> = this.componentData$.pipe(
+
+  merchandisingCarouselModel$: Observable<{
+    items$: Observable<Product>[];
+    metadata: Map<string, string>;
+  }> = this.componentData$.pipe(
     distinctUntilKeyChanged('strategy'),
-    switchMap(data => {
-      return this.cdsMerchandisingProductService.loadProductsForStrategy(
+    switchMap(data =>
+      this.cdsMerchandisingProductService.loadProductsForStrategy(
         data.strategy,
         data.numberToDisplay
+      )
+    ),
+    withLatestFrom(this.componentData$),
+    map(([merchandsingProducts, componentData]) => {
+      const metadata = this.getCarouselMetadata(
+        merchandsingProducts,
+        componentData
       );
+      merchandsingProducts.metadata = metadata;
+      return merchandsingProducts;
     }),
-    map(merchandisingProducts => merchandisingProducts.products),
-    filter(Boolean),
-    map((products: Product[]) => products.map(product => of(product)))
+    map(merchandsingProducts =>
+      this.mapMerchandisingProductsToCarouselModel(merchandsingProducts)
+    ),
+    filter<{
+      items$: Observable<Product>[];
+      metadata: Map<string, string>;
+    }>(Boolean)
   );
 
+  private getCarouselMetadata(
+    merchandisingProducts: MerchandisingProducts,
+    componentData: CmsMerchandisingCarouselComponent
+  ): Map<string, string> {
+    const metadata = new Map<string, string>();
+
+    if (merchandisingProducts.metadata) {
+      merchandisingProducts.metadata.forEach((value, name) =>
+        metadata.set(name, value)
+      );
+    }
+
+    if (
+      merchandisingProducts.products &&
+      merchandisingProducts.products.length
+    ) {
+      metadata.set('slots', merchandisingProducts.products.length.toString());
+    }
+
+    metadata.set('title', componentData.title);
+    metadata.set('name', componentData.name);
+    metadata.set('strategyid', componentData.strategy);
+    metadata.set('id', componentData.uid);
+
+    return metadata;
+  }
+
+  private mapMerchandisingProductsToCarouselModel(
+    merchandisingProducts: MerchandisingProducts
+  ): {
+    items$: Observable<Product>[];
+    metadata: Map<string, string>;
+  } {
+    return {
+      items$:
+        merchandisingProducts && merchandisingProducts.products
+          ? merchandisingProducts.products.map(product => of(product))
+          : [EMPTY],
+      metadata:
+        merchandisingProducts && merchandisingProducts.metadata
```

Comment:
As `getCarouselMetadata()` is responsible for setting the `metadata` property, this method could be renamed to something like `mapMerchandisingProductsItems()` and could deal only with `items$` property.

Reply:
Sure, updated

Reply:
The only thing that's left to do is to remove the `metadata` handling from this method